src/select.test.ts

```diff
@@ -0,0 +1,222 @@
+    const nestedSelect = q.__experimental_select({
+      'bar == "thing"': {
+        b: q.string(),
+      },
+      default: {
+        c: z.string(),
+      },
+    });
+
+    const { query, schema } = q.__experimental_select({
+      "foo > 2": nestedSelect,
+      default: {
+        baz: z.string(),
+      },
+    });
+
+    expect(query).toBe(
+      'select(foo > 2 => select(bar == "thing" => { b }, { c }), { baz })'
+    );
+    expectTypeOf({} as z.infer<typeof schema>).toEqualTypeOf<
+      { b: string } | { c: string } | { baz: string }
+    >();
+  });
+});
+
+describe("EntityQuery.select()", () => {
+  describe("with standalone select input", () => {
+    it("updates query and handles nested object types", () => {
+      const standaloneSelect = q.__experimental_select({
+        "foo > 2": {
+          bar: z.boolean(),
+        },
+        default: {
+          baz: z.string(),
+        },
+      });
+      const entityQuery = q("foo").__experimental_select(standaloneSelect);
+
+      expect(entityQuery.query).toBe(
+        "foo{...select(foo > 2 => { bar }, { baz })}"
+      );
+      expectTypeOf({} as z.infer<typeof entityQuery.schema>).toEqualTypeOf<
+        { bar: boolean } | { baz: string }
+      >();
+    });
+
+    it("converts no default condition to empty object", () => {
+      const standaloneSelect = q.__experimental_select({
+        "foo > 2": {
+          bar: z.boolean(),
+        },
+      });
+      const entityQuery = q("foo").__experimental_select(standaloneSelect);
+
+      expect(entityQuery.query).toBe("foo{...select(foo > 2 => { bar })}");
+      expectTypeOf({} as z.infer<typeof entityQuery.schema>).toEqualTypeOf<
+        { bar: boolean } | {}
+      >();
+    });
+
+    it("converts primitives to empty object", () => {
+      const standaloneSelect = q.__experimental_select({
+        "foo > 2": ["bar", z.boolean()],
+        default: {
+          baz: z.string(),
+        },
+      });
+      const entityQuery = q("foo").__experimental_select(standaloneSelect);
+
+      expect(entityQuery.query).toBe("foo{...select(foo > 2 => bar, { baz })}");
+      expectTypeOf({} as z.infer<typeof entityQuery.schema>).toEqualTypeOf<
+        {} | { baz: string }
+      >();
+    });
+
+    it("handles nested union types properly", () => {
+      const nestedSelect = q.__experimental_select({
+        "foo == 3": ["a", q.string()],
+        default: {
+          b: q.boolean(),
+        },
+      });
+
+      const standaloneSelect = q.__experimental_select({
+        "foo > 2": nestedSelect,
+        default: {
+          baz: z.string(),
+        },
+      });
+      const entityQuery = q("foo").__experimental_select(standaloneSelect);
+
+      expect(entityQuery.query).toBe(
+        "foo{...select(foo > 2 => select(foo == 3 => a, { b }), { baz })}"
+      );
+      expectTypeOf({} as z.infer<typeof entityQuery.schema>).toEqualTypeOf<
+        {} | { b: boolean } | { baz: string }
```

Comment:
Since both selects have a default, would the empty object be a possibility here? Also, I don't see the top-level `string` type for `a` accounted for.

Reply:
yea so the way groqd handles spreading on any primitive is that it just returns an empty object, so the empty object is a result of spreading on the ["a", q.string()] field and not a result of the lack of a default...

I think what I've sussed out internally that's going on in groq is that it follows these rules:

no default means null is a possible value if no conditions are met
spreading on a primitive results in an empty object

combine these two together and that's how the empty object shows up when you spread a select and there's no default condition. This assertion is basically asserting the first assumption there in the absence of the second assumption

Reply:
Funky behavior on groq's end. Wonder why they don't just make it an error to spread a primitive.
Anyway, thanks for the explanation.